integration/terraform/ec2/linux/main.tf

```diff
@@ -1,18 +1,58 @@
-resource "aws_instance" "integration-test" {
-  ami                    = data.aws_ami.latest.id
-  instance_type          = var.ec2_instance_type
-  key_name               = var.key_name
-  iam_instance_profile   = var.iam_instance_profile
-  vpc_security_group_ids = var.vpc_security_group_ids
+#####################################################################
+# Ensure there is unique testing_id for each test
+#####################################################################
+resource "random_id" "testing_id" {
+  byte_length = 8
+}
+
+#####################################################################
+# Generate EC2 Key Pair for log in access to EC2
+#####################################################################
+
+resource "tls_private_key" "ssh_key" {
+  count     = var.ssh_key == "" ? 1 : 0
+  algorithm = "RSA"
+  rsa_bits  = 4096
+}
+
+resource "aws_key_pair" "aws_ssh_key" {
+  count      = var.ssh_key == "" ? 1 : 0
+  key_name   = "ec2-key-pair-${random_id.testing_id.hex}"
+  public_key = tls_private_key.ssh_key[0].public_key_openssh
+}
+
+locals {
+  ssh_key_name        = var.ssh_key != "" ? var.key_name : aws_key_pair.aws_ssh_key[0].key_name
+  private_key_content = var.ssh_key != "" ? var.ssh_key : tls_private_key.ssh_key[0].private_key_pem
+}
+
+#####################################################################
+# Generate EC2 Instance and execute test commands
+#####################################################################
+resource "aws_instance" "cwagent" {
+  ami                         = data.aws_ami.latest.id
+  instance_type               = var.ec2_instance_type
+  key_name                    = local.ssh_key_name
+  iam_instance_profile        = aws_iam_instance_profile.cwagent_instance_profile.name
+  vpc_security_group_ids      = [aws_security_group.ec2_security_group.id]
+  associate_public_ip_address = true
+
+  tags = {
+    Name = var.test_name
+  }
+}
+
+resource "null_resource" "integration_test" {
+  # Prepare Integration Test
   provisioner "remote-exec" {
     inline = [
       "cloud-init status --wait",
       "echo clone and install agent",
-      "export PATH=$PATH:/usr/local/go/bin",
       "git clone ${var.github_repo}",
       "cd amazon-cloudwatch-agent",
       "git reset --hard ${var.github_sha}",
       "aws s3 cp s3://${var.s3_bucket}/integration-test/binary/${var.github_sha}/linux/${var.arc}/${var.binary_name} .",
+      "sleep 10",
```

Comment:
Why do we need to sleep? The `aws s3 cp` command is going to block until it completes.

Reply:
The `sleep 10` is not entirely for `aws s3 cp`, mostly it is used for avoid `"sudo ${var.install_agent}"` because of  `dpkg frontend lock is locked by another process` (https://github.com/khanhntd/amazon-cloudwatch-agent/runs/6841609674?check_suite_focus=true)

Reply:
Wasn't this part of the install_agent var before?

Reply:
Maybe I'm misunderstanding the questions but I don't think the `sleep` is part of the `install_agent` var before the changes

Reply:
@sethAmazon can you weigh in on this? I swore we accounted for the dpkg lock somewhere before previously

Reply:
Ultimately I think it's fine, though 10 seconds seems like a lot of time to wait.